pkg/operator/certrotation/target.go

```diff
@@ -117,7 +118,18 @@ func (c RotatedSelfSignedCertKeySecret) EnsureTargetCertKeyPair(ctx context.Cont
 		}
 	}
 
-	if reason := c.CertCreator.NeedNewTargetCertKeyPair(targetCertKeyPairSecret.Annotations, signingCertKeyPair, caBundleCerts, c.Refresh, c.RefreshOnlyWhenExpired); len(reason) > 0 {
+	var currentCert *x509.Certificate
+	if crtBytes, ok := targetCertKeyPairSecret.Data["tls.crt"]; ok {
+		pemCerts, err := crypto.CertsFromPEM(crtBytes)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse pem x509 tls.crt in secret %s: %w", targetCertKeyPairSecret.Name, err)
```

Comment:
should we regen the cert in this case, then?

Reply:
good question, I'm wondering if that would cause issues in etcd. Because that would also (potentially) mean that you could pre-maturely sign a leaf with a new signer certificate that's not yet rolled out in a bundle.

I'll leave this for David / Haseeb, if they have some more opinion on this.

Reply:
If you don't, you might forever crash on this and won't be able to roll out a new certificate. You'd also have to make sure that etcd would never pick the contents of the secret for all its pods because that would brick your cluster as none of the pods would be able to start.

Reply:
`pemCerts, err := crypto.CertsFromPEM(crtBytes)`

What does an error in this case represent? A malformed or corrupted `tls.crt` file right.

If so, then we really have no option but to regenerate this now otherwise this will get picked up by the installer pod in the next revision rollout. I think that's what @stlaz is referring to. (Not quite sure if the first etcd pod to rollout with the malformed peer cert will then start crashing or just report unhealthy). Plus we can't block on this and expect the secret to get patched manually with a valid tls.crt so another reason to regenerate.

> Because that would also (potentially) mean that you could pre-maturely sign a leaf with a new signer certificate that's not yet rolled out in a bundle

@tjungblu On our end, aren't we always passing the old singer from `openshift-config` into `EnsureTargetCertKeyPair()`? i.e
https://github.com/openshift/cluster-etcd-operator/blob/5b9240f586d7fc789e5a78e2608f39b729d047e0/pkg/operator/etcdcertsigner/etcdcertsignercontroller.go#L187

So even if we regenerate here we should be fine since we're using the old signer no?

And per https://github.com/openshift/cluster-etcd-operator/pull/1200 the signer in `openshift-config/etcd-singer` is only updated after we've already had a rollout with the new signer.
So the only way we prematurely regenerate here with a new signer (that isn't already bundled with each etcd member) is if we create a new signer `openshift-etcd/etcd-singer` and then immediately overwrite `openshift-config/etcd-singer` to trigger leaf-cert regen without waiting for a full rollout first.
But that would be a problem even without this case.

So I would reckon we can safely regenerate the leaf cert here. Unless I'm missing a scenario on how we end up prematurely signing the leaf cert.

Reply:
the question is: how could this certificate not become parseable anymore? it must be a user error in the first place, so is it better to alert the admin via a degraded CEO directly that the cert they created is wrong or to silent regenerate it?

Reply:
> how could this certificate not become parseable anymore? it must be a user error in the first place

That's a good point. I wasn't exactly sure on how we get to this error case. If we can agree what this state represents, (which may likely be a user overwrite)  then I can see the case for not silently recreating certs, and letting the degraded state be more obvious.

@stlaz Any thoughts on which way we want to handle this?

Reply:
Two examples pop out:
1. user malformed the secret
2. etcd memory corruption

Reply:
in neither scenarios, just recreating a certificate is the appropriate action IMHO.